### PR TITLE
LammpsInputConstructor and LammpsInputFile class [WIP]

### PR DESCRIPTION
**Pull request by manaslkht**
## Summary

1. Implementation of LammpsInputConstructor class. This class allows a line/stage wise construction of LAMMPS input file. The users can provided the LAMMPS commands and their arguments as strings or key value pairs to build the input file. This class returns LammpsInputFile object which is the object representation of LAMMPS Input file in pymatgen. 
2. The LammpsInputConstructor class introduces/implements the concept of stages/blocks in the LAMMPS input file. A stage here is defined as a block of LAMMPS input commands usually performing a specific task during the simulation such as energy minimization or NPT/NVT runs. But more broadly, a stage can also be block of LAMMPS input where simulation box is setup, a set of variables are declared or quantities are computed. 
3. Addition of LammpsInputFile. This class implements the standard I/O operation for LAMMPS input file 

## TODO

1. Allow for addition of inline comment. Currently the comments are added as a new line in the input file2. 
2. Implement methods for direct addition of LAMMPS specific commands such as adding nvt or npt runs. Suggestions are welcomed here.

### CODE REVIEW

File: pymatgen/io/lammps/inputs.py [hunk 1]
```diff
@@ -31,11 +32,111 @@
...
+        Args:
+            path: Filename to read, including path.
+
+        Returns:
+            InputFile
+        """
+        filename = path if isinstance(path, Path) else Path(path)
+        with zopen(filename, "rt") as f:
+            return LammpsInputFile.from_string(f.read())
+
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: Since you're inheriting from `InputFile`, why do you need to overload `from_file`? The contents look identical to `super().from_file()`.

Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: Since you're inheriting from `InputFile`, why do you need to overload `from_file`? The contents look identical to `super().from_file()`.

File: pymatgen/io/lammps/inputs.py [hunk 2]
```diff
@@ -271,3 +372,204 @@ def write_lammps_inputs(
...
+        """
+
+        if stage_name is None:
+            stage_name = self.curr_stage_name if self.curr_stage_name is not None else "stage %s" % self.nstages
+
+        if not self._stage_name_exist(stage_name):
+            self.input_settings[stage_name] = OrderedDict()
+            self.curr_stage_name = stage_name
+
+    def init_stage(self):
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: Any reason why this is called `init_stage` and not `add_stage`? Since the other methods start with 'add' I feel like `add_stage` would be more consistent and easier to understand

File: pymatgen/io/lammps/inputs.py [hunk 3]
```diff
@@ -271,3 +372,204 @@ def write_lammps_inputs(
...
+        Args:
+            dict_name: Dictionary with LAMMPS commands and arguments
+            stage_name: If specified, the commands are added understand
+                the specified stage else latest stage is assumed.
+        """
+
+        self._add_stage_name(stage_name)
+        for command, args in dict_name.items():
+            self.input_settings[self.curr_stage_name][command] = args
+
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: Instead of having these as separate methods, can you just do some type checking within `add_command`? Basically make the type of the the `command` arg in `add_command` `Union[str, dict, list]` and parse accordingly. The type of the `args` kwarg would then be `Optional[str]`, defaulting to None.

Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: Instead of having these as separate methods, can you just do some type checking within `add_command`? Basically make the type of the the `command` arg in `add_command` `Union[str, dict, list]` and parse accordingly. The type of the `args` kwarg would then be `Optional[str]`, defaulting to None.

File: pymatgen/io/lammps/inputs.py [hunk 4]
```diff
@@ -271,3 +372,204 @@ def write_lammps_inputs(
...
+    def init_stage(self):
+        """
+         Use this to initiate new stage/black in LAMMPS input file.
+        """
+
+        self.nstages += 1
+
+        stage_name = "stage %s" % self.nstages
+        self._add_stage_name(stage_name)
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: It seems to me that this functionality could be added to the `add_stage` command, and this method removed. You could make the `stage_commands` kwarg of `add_stage` optional. If supplied, those commands would be added, but if `None`, then `add_stage` would essentially do the same thing as `init_stage` does now.

Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: It seems to me that this functionality could be added to the `add_stage` command, and this method removed. You could make the `stage_commands` kwarg of `add_stage` optional. If supplied, those commands would be added, but if `None`, then `add_stage` would essentially do the same thing as `init_stage` does now.

File: pymatgen/io/lammps/inputs.py [hunk 5]
```diff
@@ -271,3 +372,204 @@ def write_lammps_inputs(
...
+
+    def add_command(self, command: str, args: str, stage_name=None):
+        """
+        Adds a single LAMMPS command and its arguments to LAMMPS input file.
+
+        Args:
+            command: LAMMPS command
+            args: Arguments for the LAMMPS command
+            stage_name: If a stage name is mentioned, the command is added
+                under that stage block else latest stage is assumed.
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: Maybe you can add an extra kwarg for inline comments here, e.g. 'inline_comment: Optional[str] = None'? If present, the comment would get added to the line.

File: pymatgen/io/lammps/inputs.py [hunk 6]
```diff
@@ -31,11 +32,111 @@
...
+
+        Args:
+            s: LammpsInputFile string.
+
+        Returns:
+            LammpsInputFile object
+        """
+        constr = LammpsInputConstructor()
+        return constr._from_string(s)
+
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: If you merge `LammpsInputConstructor` into this class, `from_string` can absorb the `_from_string()` and `generate_lammps_input()` methods.

File: pymatgen/io/lammps/inputs.py [hunk 7]
```diff
@@ -31,11 +32,111 @@
...
+    @staticmethod
+    def add_string(string):
+        """
+        Helper method for adding string to lammps input file.
+        Args:
+            string: String to be added to lammps input file.
+        """
+        # return '\n' + string + '\n'
+        return string + '\n'
+
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: It's odd to me that `add_string` is in `LammpsInputFile` while `add_command` and `add_comment` are in `LammpsInputConstructor`. If you are able to merge the two classes this will be more consistent (although it's not clear to me why you need `add_string` if you have the other two)

File: pymatgen/io/lammps/inputs.py [hunk 8]
```diff
@@ -31,11 +32,111 @@
...
+                if command.startswith('comment') or command.startswith('header'):
+                    if command.startswith('header'):
+                        lammps_input += '\n'
+                    lammps_input += self.add_string(args)
+                else:
+                    lammps_input += self.format_command(command, args)
+        return lammps_input
+
+    @staticmethod
+    def format_command(command, args):
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py]: I think this should be a private method, unless you see a use case for a user invoking it directly. Alternatively since it's just a one liner, why not add it directly to the body of `get_string` at line 57?

**Comment by rkingsbury** [comment]
Thanks for putting this together @manaslkht ! I've gone through and made some comments, organized into big picture and code quality / technical comments.

Please don't hesitate to ask questions if anything is unclear or your get stuck. I look forward to seeing how this evolves.

## Big picture comments
- Overall in terms of what you're trying to accomplish, I think this looks good and will be valuable. The line-by-line paradigm for building LAMMPS input files makes sense given that LAMMPS is sort of like a programming language and I would guess that many users think about their simulations this way
- As far as the interface, my main feedback is to use _fewer methods_ and _more kwargs_ . You will see this in my specific comments. For example, I think `add_command_from_string` and `add_command_from_dict` can be consolidated under `add_command` if you add some type checking to the body of the function. Remember that the more methods there are, the more new things a user has to learn and the more potential there is for similar methods to have inconsistent behavior. So it's important to keep things as simple as possible.
- Now that I've looked at this in detail, I also think there is redundancy in making `LammpsInputConstructor` separate from `LammpsInputFile`. I believe you can accomplish what you want to accomplish via just the `LammpsInputFile` class, which will make the code simpler to learn and maintain. For example. `LammpsInputConstructor._from_string()`, `LammpsInputConstructor.generate_lammps_input()` could both be absorbed into `LammpsInputFile.from_string()`. There is also no reason that you can't add attributes like `nstages` and methods like `add_command` directly to `LammpsInputFile`, in fact, this kind of extensibility is what the `InputFile` class is designed for. Can you try merging the two? 

## Code style and quality comments
- You've done a pretty good job documenting all your methods, thank you! That makes a huge difference not only to usability but even during code review.
- Please make sure your code passes linting. You can review the [pymatgen contributing guidelines](https://pymatgen.org/contributing.html?highlight=pre%20commit#direct-contributions-to-pymatgen-main-distribution) (especially item 7) for a refresher on how to set up the  linters. Specifically, I'd recomment you install the pre-commit hooks which will not allow you to commit code unless all the linters and tests pass.
- Please include type annotations (type hints) for all args/kwargs. You have them for most but there are a few without.
- When you're ready to commit, the sequence I'd recommend is 1) format your file using `black`, then check linting with 2) `flake8`, 3) `pycodestyle`, and 4) `mypy`. `mypy` in particular will be helpful for making sure you are following correct inheritance patterns and considering all edge cases. Mypy errors can be confusing to troubleshoot so please let us know if you get stuck.
- Definitely add unit tests. These should go in `pymatgen.io.lammps.tests.test_inputs.py` The existing tests in that file use the `unittest` framework but for newer tests we are encouraging `pytest` style tests which are a bit simpler to write. You can refer to these test files as examples: [example1](https://github.com/materialsproject/maggma/blob/db6e095ab8c41e923bae856d54e6904a6b3a3293/tests/stores/test_file_store.py) [example2](https://github.com/materialsproject/pymatgen/blob/master/pymatgen/entries/tests/test_computed_entries.py) [example3](https://github.com/materialsproject/pymatgen/blob/master/pymatgen/io/tests/test_packmol.py)
- Your goal should be to have >95% test coverage for the code you add. A good way to check test coverage is to install [`coverage`](https://coverage.readthedocs.io/en/6.3.2/) and use it when you run pytest (especially when the code is nearing a final state)

**Review by rkingsbury — CHANGES_REQUESTED** [review]

File: pymatgen/io/lammps/inputs.py (hunk 2 above)
Comment (manaslkht) [reply pymatgen/io/lammps/inputs.py]: Yes, the `add_stage` method already exists in this class which is meant to add an entire stage to the lammps input. The idea behind `init_stage` is to initialize a stage such that subsequent command addition happens to the new stage. The `add_stage` actually already makes use of `init_stage` internally. Let me know if this makes sense. 

In the future, I will extend `init_stage` method such that we are able to pass the stage name which is going to "initiate" back the passed stage name such that the sequence command addition happens to that stage.

Comment (rkingsbury) [reply pymatgen/io/lammps/inputs.py]: I see. It's still not clear to me that separate methods are needed. It sounds like you want to have the ability to append information to an existing stage (hence `init_stage` + subsequent commands) vs. `add_stage` which adds a complete stage all at once.

The analogy with regular python classes (here, list) would be
```
new_list = []
list.append('stuff')
```
vs.
```
new_list = ["stuff"]
```
Am I understanding that right? If so, I feel like there may be a more "pythonic" way to accomplish this that doesn't require two separate methods.

**Comment by orionarcher** [comment]
This is a great start @manaslkht! I second all of @rkingsbury's comments here. I also think you could get by with only a `LammpsInputFile` object, which would allow for a simpler overall implementation.

File: pymatgen/io/lammps/inputs.py (hunk 1 above)
Comment (manaslkht) [reply pymatgen/io/lammps/inputs.py]: Yes, I am required to overload this method due to the inheritance problem we discussed. Please let me know if there is a better way to code this to avoid overloading inherited methods.

File: pymatgen/io/lammps/inputs.py (hunk 3 above)
Comment (manaslkht) [reply pymatgen/io/lammps/inputs.py]: I agree. I have merged all the separate methods for adding commands into one method.

File: pymatgen/io/lammps/inputs.py (hunk 5 above)
Comment (manaslkht) [reply pymatgen/io/lammps/inputs.py]: I will implement this soon.

File: pymatgen/io/lammps/inputs.py (hunk 6 above)
Comment (manaslkht) [reply pymatgen/io/lammps/inputs.py]: Yes, I have merged the two classes.

File: pymatgen/io/lammps/inputs.py (hunk 8 above)
Comment (manaslkht) [reply pymatgen/io/lammps/inputs.py]: Have implemented this change.

**Comment by manaslkht** [comment]
@rkingsbury Hi Ryan, Thank you so much for your detailed feedback. The following are the major changes that I have implemented based on your suggestions. 

1. I have merged the `LammaInputConstructor` class into the `LammpsInputFile` class as you suggested. LammpsInputFile now has all the functionalities of the formal `LammaInputConstructor` class. 
2. I have merged all the add command methods into to one method that accepts `string`, `list` or `dict` of commands. 

I am still working on the problem of avoiding overloading some of the inherited class methods such as `from_string`. In my code I am using other methods of the class to process the string and return the LammsInputFile object. So the reference to "self" is necessary. Your inputs here will be necessary. 

I will finish writing the unit tests for these methods soon and push it. In the meantime please review the current code and let me know if you would like to make more changes. 

Thank you!

File: pymatgen/io/lammps/inputs.py [hunk 9]
```diff
@@ -9,18 +9,18 @@
...
-from monty.json import MSONable
 from monty.io import zopen
+from monty.json import MSONable
 
 from pymatgen.io.core import InputFile
-from pymatgen.io.template import TemplateInputGen
 from pymatgen.io.lammps.data import LammpsData  # , CombinedData
+from pymatgen.io.template import TemplateInputGen
 
 __author__ = "Kiran Mathew, Brandon Wood, Zhi Deng"
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:25]: Feel free to add your name to the author list!

File: pymatgen/io/lammps/inputs.py [hunk 10]
```diff
@@ -33,10 +33,114 @@
...
     def __init__(self, input_settings: OrderedDict = None):
-        self.input_settings = input_settings
+        """
+        Args:
+            input_settings: Ordered Dictionary of LAMMPS input settings.
+        """
+
+        self.nstages = 0
+        self.ncomments = 0
+        self.input_settings = input_settings if input_settings is not None else OrderedDict()
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:54]: I believe you can omit the `is not None` here.

File: pymatgen/io/lammps/inputs.py [hunk 11]
```diff
@@ -33,10 +33,114 @@
...
+        self.curr_stage_name = None
+
+    def init_stage(self):
+        """
+        Use this to initiate new stage/black in LAMMPS input file.
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:59]: "black" -> "block"

File: pymatgen/io/lammps/inputs.py [hunk 12]
```diff
@@ -33,10 +33,114 @@
...
+        """
+
+        self.nstages += 1
+
+        stage_name = "stage %s" % self.nstages
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:64]: Minor comment - here you use old style format strings but at line 84 you use an f-string. It's OK as-is but I'd suggest you keep your code style consistent throughout (and use f-strings everywhere since they are more modern)

File: pymatgen/io/lammps/inputs.py [hunk 13]
```diff
@@ -33,10 +33,114 @@
...
+        else:
+            self.add_comment(header, is_stage_header=True)  # type: ignore
+
+        self.add_commands(commands=stage_commands)
+
+    def generate_lammps_input(self):
+        """
+        Returns: LammpsInputFile object
+        """
+        return LammpsInputFile(self.input_settings)
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:143]: It's not clear to me what purpose this method serves. Isn't this basically returning a copy of the object itself? I think it can be removed.

File: pymatgen/io/lammps/inputs.py [hunk 14]
```diff
@@ -33,10 +33,114 @@
...
+    def add_commands(self, commands: Union[str, list, dict], stage_name: str = None):
+        """
+        Adds LAMMPS command/s and its arguments to LAMMPS input file.
+
+        Args:
+            commands: LAMMPS command/s. Can pass single string or list of LAMMPS command
+            with its arguments. Also accepts dictionary of LAMMPS commands are
+            corresponding arguments as key value pairs.
+            stage_name: If a stage name is mentioned, the command is added
+                under that stage block else latest stage is assumed.
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:95]: Minor typographical comments: I think it's more conventional to say "command(s)" rather than "command/s" to indicate possible plural. Also the syntax of "Also accepts dictionary of LAMMPS commands are
            corresponding arguments as key value pairs." is not quite right. I think you mean "commands and corresponding arguments as key,value pairs"

File: pymatgen/io/lammps/inputs.py [hunk 15]
```diff
@@ -120,31 +220,66 @@ def as_dict(self):
...
+        """
+        if self.curr_stage_name is None:
+            self.init_stage()
+        self._add_stage_name(stage_name)
+
+        if args is None:
+            string_split = command.split()
+            command = string_split[0]
+            args = " ".join(string_split[1:])
+        self.input_settings[self.curr_stage_name][command] = args
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:275]: I think there might be potential for unintended behavior here. What if I pass a `stage_name` kwarg that is different than `self.curr_stage` (which already has a value, not `None`) AND `stage_name` is already in `self.input_settings`?

If I follow the logic correctly, `add_stage_name` would not do anything in that case, and the command would be added under the key `self.curr_stage`  and not `stage_name`.

File: pymatgen/io/lammps/inputs.py [hunk 16]
```diff
@@ -46,58 +150,54 @@ def get_string(self):
...
         """
-        constr = LammpsInputConstructor()
-        return constr._from_string(s)
+        filename = path if isinstance(path, Path) else Path(path)
+        with zopen(filename, "rt") as f:
+            return self.from_string(f.read())
 
     @classmethod
-    def from_dict(cls, d):
+    def from_dict(cls, d: dict):
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:200]: My apologies if we've already discussed this, but are you sure it's necessary to have custom `as_dict` and `from_dict` methods here? `InputFile` inherits from `MSONable` which means we should get `as_dict` and `from_dict` "for free".

Is this because of the `OrderedDict` requirement?

File: pymatgen/io/lammps/inputs.py [hunk 17]
```diff
@@ -120,31 +220,66 @@ def as_dict(self):
         d["@class"] = type(self).__name__
         return d
 
+    def __str__(self):
+        return self.get_string()
+
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:225]: Can you explain why this is needed? Do you think it's something that should be a part of the base class instead of this one?

File: pymatgen/io/lammps/inputs.py [hunk 18]
```diff
@@ -46,58 +150,54 @@ def get_string(self):
...
 
         Returns:
-            LammpsInputFile object
+            InputFile
         """
-        constr = LammpsInputConstructor()
-        return constr._from_string(s)
+        filename = path if isinstance(path, Path) else Path(path)
+        with zopen(filename, "rt") as f:
+            return self.from_string(f.read())
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:197]: Again apologies if we've discussed this already, but what is it that's preventing you from using the inherited `from_file` method here? I don't see anything here that should be specific to the internal structure (i.e., `OrderedDict` of `LammpsInputFile`.

File: pymatgen/io/lammps/inputs.py [hunk 19]
```diff
@@ -46,58 +150,54 @@ def get_string(self):
...
-    @staticmethod
-    def add_string(string):
-        """
-        Helper method for adding string to lammps input file.
-        Args:
-            string: String to be added to lammps input file.
-        """
-        # return '\n' + string + '\n'
-        return string + '\n'
+        return LammpsInputFile(self.input_settings)
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:183]: OK, this method cannot reference `self` because it's purpose is to _construct_ a `LammpsInputFile` from a string. So there is no `self` until this method returns it. Right now your method is written as if a `LammpsInputFile` already exists and you want to update it.

I see two approaches you could take
1. Create an empty `LammpsInputFile`, then `add_comment` / `add_commands` to update it, and return that object
2. Use string parsing to build an `OrderedDict`, then init your `LammpsInputFile` at the end and return that

If you use Option 2, you would need to make `_clean_lines`, `add_comment`, and `add_commands` static methods. That makes sense for `clean_lines` b/c it's just a utility function that has nothing to do with the contents of this class, but `add_comment` and `add_commands` make less sense as static methods.

File: pymatgen/io/lammps/inputs.py [hunk 20]
```diff
@@ -46,58 +150,54 @@ def get_string(self):
 
         """
 
-        lammps_input = '## LAMMPS input generated from LammpsInputConstructor\n\n'
+        lammps_input = "## LAMMPS input generated from LammpsInputConstructor\n\n"
```
Comment (rkingsbury) [inline pymatgen/io/lammps/inputs.py:153]: "LammpsInputConstructor" -> "LammpsInputFile"

File: pymatgen/io/lammps/inputs.py (hunk 11 above)
Comment (rkingsbury) [reply pymatgen/io/lammps/inputs.py:59]: I still feel that his method could be combined with `add_stage`. All you would have to do is make `stage_commands` optional and then with default kwargs `add_stage` would just create an empty stage.

I think it will be difficult to communicate the difference between `init_stage` and `add_stage` to users. I know we've discussed this before, and if there is a strong technical reason to separate them you can certainly convince me that it should be so. But I'll need a refresher on why it's necessary :)

**Review by rkingsbury — CHANGES_REQUESTED** [review]
Solid progress! I've made a series of specific comments; let's discuss those before we dive into the weeds of the outstanding `mypy` issues.

My only general comment is related to the semantics of your interface. `add_stage` and `add_comment` are singular and add one item at a time, but `add_commands` is plural and designed to take in a single or multiple commands at a time. As a user, I would expect all 3 methods to work the same way because they all start with `add_` and involve adding something to the object.

I can appreciate that it's more likely to want to add multiple commands in a batch compared to comments or stages, so it's nice to have the ability to pass a list. However, I can guarantee that a lot of users are going to think the method is `add_command` and will type that a few times before remembering _just that one_ method is plural.

What I suggest is to put all of the important functionality to process a single command in `add_command`, and then define a lightweight `add_commands` that calls `add_command` in a for loop. That way you provide a consistent interface with singular verbs but also give the option to process commands in batch. We've done for example in the `Compatibility` classes where we have a `process_entry` method that is not really designed to be used by end users, and then a `process_entries` method that calls it.

**Comment by orionarcher** [comment]
Looking good Manas! Regarding Ryan's comment:

> What I suggest is to put all of the important functionality to process a single command in add_command, and then define a lightweight add_commands that calls add_command in a for loop. That way you provide a consistent interface with singular verbs but also give the option to process commands in batch. We've done for example in the Compatibility classes where we have a process_entry method that is not really designed to be used by end users, and then a process_entries method that calls it.

Another way to approach this, (that could make the api less verbose) would be to let add_command take either one or multiple commands. If the input is a string, it will add a single item, if the element is a list, it will add multiple. This would mean you don't have to have two separate methods and the user can use it however feels natural.
